torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1803,6 +1853,10 @@ def _pre_backward_hook(*unused: Any) -> None:
                 # pre-backward hook.
                 torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
 
+            # Start of a backward pass for the first time in an backward pass.
```

Comment:
I moved this state transition to earlier in the pre-backward hook since otherwise the call to `_rebuild_full_params()` is still under `IDLE` rather than `PRE_BACKWARD`. The training state should not affect the execution of any of the code between this new position and the old position a few lines down. However, it could be good to get a second pair of eyes to sanity check this.

Reply:
Hmmm, I guess it depends on whether we want to consider this all_gather as part of backward_pre or not?